Changing ReferenceGrant KEP to broader Referential Authorization KEP (#4387)
kubernetes/enhancements: keps/sig-auth/3766-referential-authorization/README.md

Reviewer: Suggestion for the Alternatives section:
```diff
 ## Alternatives
+
-1. ReferenceGrant could remain as a CRD
+## Implement a revised ReferenceGrant as a CustomResourceDefinition
+
-This would probably be fine, we just don't really have a good place for it to
-live. This could also complicate installation of Gateway API and other APIs that
-depended on this.
+Implement the API defined within this KEP, but as a CustomResourceDefinition rather than
+as an in-tree API.
+A CRD implementation is feasible but would make integration with in-tree authorization features,
+such as Kubernetes RBAC, more complex. Cluster operators who wanted to adopt an out-of-tree
+ReferenceGrant would need to deploy a webhook or other mechanism outside the core control
+plane to implement the access mechanism, and ensuring that this mechanism operates reliably
+then falls on each cluster and its operators.
+
+Additionally, anything that integrated with ReferenceGrant as a CRD would need to be prepared
+to handle the case where existing ReferenceGrants and the entire API could be removed from a
+cluster. With the in-tree option, we get a point in time where every supported Kubernetes release
+provides ReferenceGrant as a stable API, and other software can assume it will remain available.
+
-2. Every API that wanted to support cross-namespace references could maintain their own version of ReferenceGrant
+### Every API that wants to support cross-namespace references maintains its own equivalent to ReferenceGrant

 This would be a confusing mess, we should avoid this at all costs.
+
+### Use CEL to define ClusterReferencePattern, rather than JSONPath
+
+ [details go here, or use an UNRESOLVED block]
```

Author: Updated, thanks!

  within Gateway API.
* In addition to ReferenceGrant's capabilities, expand the scope of this KEP so
  that it enables:
  * Authorizing controllers to read only resources that have been referenced by
    resources they are implementing.
  * Building a more generic approach that will enable generic controllers and
    authorizers to implement this API without prior knowledge of resources that
    references are from or to.
* Clearly define how ReferenceGrant should be used, including both current use
  cases and guidance for future use cases.
* Implement a library to ensure that ReferenceGrant is implemented consistently
  by all controllers.

### Non-Goals

* Advanced capabilities like label selectors and the ability to allow references
  from any namespace are likely reasonable future extensions of this API but are
  intentionally out of scope for the initial work.

## Proposal

Create 3 new API types in a new `reference.authorization.k8s.io` API Group:

1. **ClusterReferenceGrant:** This resource grants access to resources by
   following references within the same namespace. For example, a `gateways`
   ClusterReferenceGrant would grant access to Gateway controllers for

>    same-namespace references from Gateways to TLS Secrets. API developers would

Reviewer 2: I still think that we should separate the API object for defining how the API type is shaped (versions, JSONPaths/CELExprs, etc.) and allowing "within namespace references", and not bake them into the same object.

This way one could avoid allowing all within-namespace references, relying only on "individual" ReferenceGrants defined in the cluster (for specific within and cross-namespace references).

that class.
Proposed fields from GEP-91 are added to the v1 version, but not v1beta1.

```yaml
kind: ClusterReferenceGrant
apiVersion: reference.authorization.k8s.io/v1alpha1
metadata:
  name: gateways
from:
  group: gateway.networking.k8s.io
  resource: gateways
versions:
  - version: v1
    classPath: ".spec.gatewayClassName"
    references:
      - path: "$.spec.listeners[*].tls.certificateRefs[[?(@.group=='' && @.kind=='Secret')].name"
        to:
          group: ""
          resource: secrets
        for: tls-serving
      - path: "$.spec.listeners[*].tls.clientValidation.caCertificateRefs[[?(@.group=='' && @.kind=='Secret')].name"
        to:
          group: ""
          resource: secrets
        for: tls-client-validation
      - path: "$.spec.listeners[*].tls.clientValidation.caCertificateRefs[[?(@.group=='' && @.kind=='ConfigMap')].name"
        to:
          group: ""
          resource: configmaps
        for: tls-client-validation
  - version: v1beta1
    classPath: ".spec.gatewayClassName"
    references:
      - path: "$.spec.listeners[*].tls.certificateRefs[[?(@.group=='' && @.kind=='Secret')].name"

>         # Kubernetes JSONPath doesn't actually support this kind of compound boolean filter expression

Reviewer 3: This is more explicit and more annoying than trying to build some sort of duck-typing behavior.
It does make it easier to disambiguate any use of `kind` which is preferable: [API Conventions: kind vs. resource](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#kind-vs-resource)

We still should probably extract namespace here when possible, even if it's just to rule out cross-namespace references. We could access the namespace of the root level object in JSONPath or CEL or something.

Maybe if the path returns a struct of {name, namespace}, we can use it and compare?
Do we force API owners to somehow explicitly declare references are same-namespace?

Do we hold onto cross-namespace references that are declared here as well and try to match those up with any ReferenceGrants from those namespaces?

Presumably, we always ignore {group, resource, kind} in the struct because it must be opt-in.
We can't always know the shape of the API we are descending into.
... unless we can add some sort of safe, sensible matching behavior that automatically checks for the most canonical kinds of object references?

Reviewer 4: There might be APIs that references just a named resource without specifying its `kind` explicitly in the reference. I was thinking that the output of the CEL/JSONPath expression can be something like `kind/ns/name` so the authorizer would know what referenced resource it should look for.
Unless I am missing something and this design already account for such references

  for a ClusterReferenceConsumer that lists no classNames? Software authors, API
  owners, and cluster admins all own different pieces of the config here. An 
  optional field that fails open can easily be missed by a cluster admin. An  
  opt-in wildcard can become canonical in practice.
* Adding classNames burdens cluster admins with synchronizing their classNames 
  between their IngressClasses, GatewayClasses, StorageClasses, etc.
  IngressClass and GatewayClass already have a `spec.controller` and
  `spec.controllerName` field. StorageClass has a `provisioner` field.
  There is an opportunity here to have ClusterReferenceConsumers select into
  these class types by their controller/provisioner name.
  It requires API owners to define how to extract which of classNames pair up 
  with which controller/provisioner names.
  The benefit is that ClusterReferenceConsumers can specify a more canonical
  controllerName field, leaving all of the declarations of classNames contained
  within their specialized IngressClass/GatewayClass/StorageClass API.
  This extra layer of indirection reduces config toil and duplication, but is

>   arguably more incomprehensible.

Reviewer 3: I've currently proposed just extracting classNames from referencing objects and matching them to consumers.
Then cluster-admins have to ensure classNames are correct for a consumer if they change them.
( maybe this get's templated by a helm-chart/operator/packageinstall ? )

I'm curious for feedback for this other idea of having another API definition mapping `spec.controller` and `provisioner` to a `ClusterReferenceConsumer.controllerName` so we can match GatewayClasses and IngressClasses to their Subject. It could live inside of ClusterReferenceGrant, or it could be it's own thing like ClassControllerMap:

```
kind: ClassControllerMap
apiVersion: reference.authorization.k8s.io/v1alpha1
metadata:
  name: gatewayclass
spec:
  versions:
  - version: v1
    group: gateway.networking.k8s.io/v1
    resource: gatewayclass
    controllerPath: .spec.controllerName
```

compromise an Ingress or Gateway controller could potentially provide access to
all Secrets in a cluster. Ideally, a solution that is related to referential
authorization could also help mitigate this problem.

### Goals

* Introduce API(s) that enable the same capabilities that ReferenceGrant does
  within Gateway API.
* In addition to ReferenceGrant's capabilities, expand the scope of this KEP so
  that it enables:
  * Authorizing controllers to read only resources that have been referenced by
    resources they are implementing.
  * Building a more generic approach that will enable generic controllers and
    authorizers to implement this API without prior knowledge of resources that
    references are from or to.
* Clearly define how ReferenceGrant should be used, including both current use
  cases and guidance for future use cases.
* Implement a library to ensure that ReferenceGrant is implemented consistently
  by all controllers.

### Non-Goals

* Advanced capabilities like label selectors and the ability to allow references
  from any namespace are likely reasonable future extensions of this API but are
  intentionally out of scope for the initial work.

## Proposal

> Create 3 new resources in a new `reference.authorization.k8s.io` API Group:

Author: Updated, thanks!

1. **ClusterReferencePattern:** This resource defines a common reference
   pattern. For example, a `gateway-tls` ClusterReferencePattern would define
   how references from Gateways to TLS Secrets work. Authors of APIs would
   bundle these definitions with their APIs.
2. **ClusterReferenceConsumer:** This resource links a consumer (defined by a
   RBAC Subject) to one or more ClusterReferencePatterns. For example, a Gateway
   controller that used a ServiceAccount for auth, would include a
   ClusterReferenceConsumer tying that ServiceAccount to the `gateway-tls`
   pattern described above.
3. **ReferenceGrant:** This resource authorizes a specific instance of a
   ClusterReferencePattern. For example, if an owner of a Secret in namespace
   `foo` wants to allow a TLS reference to that Secret from a Gateway in

>    namespace `bar`, they would create a `ReferenceGrant` that allowed the

Author: Updated, thanks!

sig-storage](https://kubernetes.io/blog/2023/01/02/cross-namespace-data-sources-alpha/)
to enable cross-namespace data sources.

This KEP proposes building that overall idea of referential authorization
directly into Kubernetes. This will build on the original idea of ReferenceGrant
but add additional capabilities along the way.

## Motivation

Any project that wants to enable cross-namespace references currently has to
choose between introducing a dependency on Gateway API's ReferenceGrant or
creating a new API that would be partially redundant (leading to confusion for
users).

> Recent interest between SIGs has made it clear that ReferenceGrant is wanted for

Reviewer 2: Could you mention which SIGs have been interested in this approach? Storage IIRC, maybe network?

Reviewer 3: Flux also intends to consume this API.

https://github.com/fluxcd/flux2/pull/2092#issuecomment-1479788234

Author: Added, thanks!

use cases other than Gateway API. We would like to move ReferenceGrant to a
neutral home (ideally, under SIG Auth) in order to make it the canonical API for
managing references across namespaces.

> At the same time, Ingress and Gateway controllers are often deployed with read
> access to all Secrets in a cluster. This can be a significant security risk as

Reviewer 2: ```suggestion
Due to lack of better alternatives, Ingress and Gateway controllers, for instance, are often deployed
with read access to all Secrets in the cluster. This can be a significant security risk as
```

Author: Updated, thanks!

these components often already sit on the edge of a cluster. Finding a way to
compromise an Ingress or Gateway controller could potentially provide access to

> all Secrets in a cluster. Ideally, a solution that is related to referential
> authorization could also help mitigate this problem.

Reviewer 2: ```suggestion
all Secrets in a cluster. This KEP provides a more fine-grained approach to
authorizing cross- and in-namespace references, such that users have the
realistic possibility of removing coarse RBAC rules such as that the Ingress
controller can access all the Secrets in the cluster. This KEP does not offer
extra security for references, in case RBAC rules are applied in such a
coarse manner.
```

Author: Updated, thanks!

### Goals

* Introduce API(s) that enable the same capabilities that ReferenceGrant does
  within Gateway API.
* In addition to ReferenceGrant's capabilities, expand the scope of this KEP so
  that it enables:
  * Authorizing controllers to read only resources that have been referenced by
    resources they are implementing.
  * Building a more generic approach that will enable generic controllers and
    authorizers to implement this API without prior knowledge of resources that
    references are from or to.
* Clearly define how ReferenceGrant should be used, including both current use
  cases and guidance for future use cases.
* Implement a library to ensure that ReferenceGrant is implemented consistently
  by all controllers.

### Non-Goals

* Advanced capabilities like label selectors and the ability to allow references
  from any namespace are likely reasonable future extensions of this API but are
  intentionally out of scope for the initial work.

## Proposal

Create 3 new API types in a new `reference.authorization.k8s.io` API Group:

> 1. **ClusterReferenceGrant:** This resource grants access to resources by

Reviewer 2: I think it would be interesting to show the example where in kube today we implicitely allow Pods to reference Secrets/ConfigMaps/etc. within the same namespace with no extra authorization.

To get a clearer picture, it'd be quite informational to go through the k8s API and find all these references and document them. I guess in the future, kube itself could create default ClusterReferenceGrant objects which are deletable for users that don't want this behavior out of the box.

I think it can be confusing to users to find out that for some resources (Pods), they don't need a ClusterReferenceGrant to access the same-namespace Secret, but for some resources (HTTPRoute), they do.

> 1. **ClusterReferenceGrant:** This resource grants access to resources by

Reviewer 5: Naming-wise, if this is "same namespace," is it confusing to call it "Cluster"ReferenceGrant, given that ClusterRole/ClusterRoleBinding are cluster-wide instead of same namespace?

Reviewer 5: Though I guess the behavior here is a single subject is still granted permission on something across multiple namespaces, so perhaps that makes sense. The "same-namespace" bit is just referring to the references between namespaced resources, not access paths from subjects to namespaces.

Author: Yeah it's strange to find the right name here, I was trying to go with the idea that this resource is very similar to ReferenceGrant but is cluster-scoped. Would be very open to a different name here, it could potentially make it easier to understand. Anyone have ideas here? We can also revert to ClusterReferencePattern which may be more clear.

cc @luxas @stealthybox @ttakahashi21

Reviewer 2: Yes, this is something that confused me as well, I think we somehow have to rename it.

WithinNamespaceReferenceGrant is a mouthful, but kind of describes it well
SameNamespaceReferenceGrant
NamespacedReferenceGrant is confusing, as ReferenceGrant is a namespaced resource already :D 

somehow I'd like to get in the "namespace" in this name, although that makes the resource name quite long.
but important would be to emphasise that ReferenceGrant is fine-grained, SameNamespaceReferenceGrant (or similar) is a little more coarse, or more like a "generic rule"

Reviewer 6: I agree with the name change. Certainly, I think the name should convey the meaning that it is for within the same namespace.
How about @stealthybox ?

   include a ClusterReferenceConsumer tying that ServiceAccount to references
   from "Gateways" to "Secrets" for "tls". This would link that consumer to any
   (Cluster)ReferenceGrants with the same `from`, `to`, and `for` values.
3. **ReferenceGrant:** This resource authorizes specific instances of
   cross-namespace references. For example, if an owner of a `acme-cert` Secret
   in namespace `foo` wants to allow a TLS reference to that Secret from a
   Gateway in namespace `bar`, they would create a ReferenceGrant that allowed
   references from Gateways in the `bar` namespace to the `acme-cert` Secret in
   their namespace (`foo`).

### Risks and Mitigations

#### Potential for Variations Among Implementations
Although the authorization logic for ReferenceGrant will be built into
Kubernetes, controllers will still need to perform additional checks. For
example, consider the case where 2 Gateways target the same Secret, but only
one of those references is allowed by a ReferenceGrant. The authorization logic
built into Kubernetes would have granted the controller access to the Secret
but the controller would still need to perform separate checks to ensure that

> a ReferenceGrant had allowed a specific reference.

Reviewer 5: Is there any way a controller could indicate in an request that it's following a reference (e.g. a "from" resource?), such that the authorizer could check that there is in fact a grant from that referrer resource to the target resource?

Author: That's definitely something that @deads2k and @liggitt have suggested, but it's unclear how that would work. I think it was @stealthybox that suggested something like impersonation here but that turned into a dead end IIRC. IMO this is something that could be added after the fact without actually changing the user-facing API. For now I think the current proposal solves ~90% of the problem and still leaves room for solving this part in the future (with the same API).

sig-storage](https://kubernetes.io/blog/2023/01/02/cross-namespace-data-sources-alpha/)
to enable cross-namespace data sources.

This KEP proposes building that overall idea of referential authorization
directly into Kubernetes. This will build on the original idea of ReferenceGrant
but add additional capabilities along the way.

> ## Diagrams

Author: In case it's too annoying to view these in GitHub, here are some mermaid links:

* [Without RBAC](https://www.mermaidchart.com/raw/f0ec5f4d-b7e3-45b4-bc54-e9c95fa3548f?theme=light&version=v0.1&format=svg) 
* [With RBAC](https://www.mermaidchart.com/raw/268fb055-78ed-4cd6-9bd2-e3f2f7df448b?theme=light&version=v0.1&format=svg)

When thinking about ReferenceGrant, it is important to remember that it does not
do anything by itself. It *Grants* the *possibility* of making a *Reference*
across namespaces. It's intended that _another object_ (that is, the From
object) complete the handshake by creating a reference to the referent object

> (the To object).

Author: Updated, thanks!

#### ReferenceGrant authors must have sufficient access

> Anyone creating or updating a ReferenceGrant MUST have at least the level of access

Author: Updated, thanks!

to the resources they are providing access to. For example, if the
ClusterReferencePattern specifies `[get, list, watch]` as `verbs`, the author of
any ReferenceGrant using that pattern must also have access for all of those
verbs. If that authorization check fails, the ReferenceGrant update or create
action will also fail.

> #### `Resource` vs `Kind`

Author: Updated, thanks!

resource (the `to` resource).

#### ReferenceGrant authors must have sufficient access

Anyone creating or updating a ReferenceGrant **MUST** have at least the level of
access to the resources they are providing access to. For the scope of this KEP,
referential authorization translates to `[get, list, watch]` access. In the
future, that may be configurable.

#### `resource` versus `kind`

When creating a metaresource (that is, a resource that targets other resources)
like ReferenceGrant, it's important to consider if the metaresource uses the more
common `kind` or the more correct `resource`.

In the original Gateway API implementation, we chose to use `kind` rather than
`resource`, mainly to improve the user experience. That is, it's easier users
to take the value from the `kind` field at the top of the YAML they are already
using, and put it straight into these fields, rather than needing to do a
kind-resource lookup for every user's interaction with the API. @robscott even
ended up making https://github.com/kubernetes/community/pull/5973 to clarify
the API conventions.

However, in discussion on this KEP, it's clear that the more generic nature of

> _this_ API requires the additional specificity that `resource` provides. The new
> version communicates the scope more clearly because `group`+`resource` is

Reviewer 2: ```suggestion
_this_ API requires the additional specificity that `resource` provides. The design
in this KEP communicates the scope more clearly because `group`+`resource` is
```

Reviewer 2: removed "version" in the text here as that is also a field, which can be confusing when reading.

Author: Updated, thanks!

the API conventions.

However, in discussion on this KEP, it's clear that the more generic nature of

> _this_ API requires the additional specificity that `Resource` provides. The new

Author: Updated, thanks!

Create 3 new API types in a new `reference.authorization.k8s.io` API Group:

1. **ClusterReferenceGrant:** This resource grants access to resources by

>    following references within the same namespace. For example, a `gateways`

Reviewer 3: Ah fair!
I suppose an API author can still split up a resource's references by purpose.
The new VersionedReferencePaths allows for all paths/purposes to be listed for a single resource -- this in my head maps 1:1 to a CustomResourceDefinition, but there's also nothing stopping you from splitting it out.

#### Revocation behavior

When a ReferenceGrant is deleted/revoked, the corresponding authorization for
controllers to access that resource via that ReferenceGrant will also be
revoked.

Unfortunately that only covers a portion of the behavior that ReferenceGrant
enables. APIs that make use of ReferenceGrant will need to clearly define how
the revocation of a ReferenceGrant works. With that said, we expect the

> following guidelines to be rules to apply to ALL implementations of the API:

Author: Updated, thanks!

  forwarding requests or persisting data) cannot be undone, but no future
  actions should be allowed.

### Example Usage

#### Gateway API Cross-Namespace Secret Reference

Authors of APIs can bundle ClusterReferencePatterns with their API definitions
for any references they expect to be common. For example, Gateway API could
bundle the following ClusterReferencePattern to describe how Gateways reference

> TLS Secrets. Note that this also allows API authors to define a `SameNamespace`

Author: Updated, thanks!

> `SameNamespace` `baselineGrant` that would automatically authorize all
> references following this pattern without crossing namespace boundaries.

Reviewer: (for clarity)
```suggestion
`baselineGrant: SameNamespace`; setting the `baselineGrant` of a ClusterReferencePattern
to `SameNamespace` would automatically authorize all references that match the
ClusterReferencePattern and that do not cross namespace boundaries.
```

(this is more like how we'd write it in the tech docs; I'm recommending it here because I actually thought
there might be a BaselineGrant API that I couldn't see).

```yaml
kind: ClusterReferencePattern
apiVersion: reference.authorization.k8s.io/v1alpha1
metadata:

>   name: gateway-tls

Author: Mentioned this in Slack, but @liggitt also had very similar feedback here which led to an overhaul here, moving away from this resource requiring meaningful/reserved names.

group: gateway.networking.k8s.io
resource: gateways
version: v1

> path: ".spec.listeners[*].tls.certificateRefs[*]"

Reviewer 7: What is the schema contained in `certificateRefs`?  Name, namespace,name, group/resource/namespace/name?  Apologies if this is somewhere outside the diff

Author: This is actually a big open question with the updated design. Currently we specify that each reference needs to be `To` a specific group and resource to be authorized by a (Cluster)ReferenceGrant. One potential interpretation of that is that we would just take the name(s) at this JSON path and assume they were to the group and resource specified by the (Cluster)ReferenceGrant. That does not feel particularly safe though, so another option would be to ensure that the Group and Resource specified by the (Cluster)ReferenceGrant also aligned with the reference here.

Author: cc @liggitt @enj @stealthybox @luxas

Reviewer 3: We can implement a sort of duck-typing behavior where is the `path` is:
- a string representing name in the same namespace
- an object that has a name field but no namespace field (assume same namespace)
- an object that has a name and namespace field (if empty, assume same namespace)

This feels like a decent way to support
- Ingress API's use of `secretName`
- Pod style `secretRef` and `envFrom` API's
- API's that have cross-object references

It also satisfies Flux's current API's which don't specify a group for cross-namespace refs -- just and object with name, namespace, version and kind.

Another sensible thing to do would be to also attempt to parse any `group`/`resource`/`kind` present in the object at the path.

Since every grant, whether it's cluster-level or namespaced, specifies a from+to+for, it will usually be unambiguous whether a reference to a specific resource for a specific purpose is allowed.
The edge case is if an API makes a reference where the resource group or kind it's operating on can be changed in a different part of the API spec. Then potentially, you could have ClusterReferenceGrants listing and granting names for group+version+resource that isn't intended. (This isn't much of an issue with namespaced ReferenceGrants because they list specific names already)

One way to support this style of strange API would be to add some sort of compound condition to the ClusterReferenceGrant `path`  that lets you say "and this other field must equal this String".
ex: `config.kind == ConfigMap` or `sourceRef.kind == GitRepository`
This would allow API owners to make more specific ClusterReferenceGrants.
It also opens a use-case for things like `gatewayClass == Contour`. (whether we want to encourage that is another question)
This might already be possible if we use `path` with some sort of compound CEL expression?

The main goal of this very complex heuristic behavior is to support many different types of API's.
The alternative would be to add fields to ClusterReferenceGrant that tell it what type it should expect at the path and how to parse or assume the needed information:
- name
- namespace
- group
- resource / kind

Reviewer: @stealthybox: sounds like good grounds for a thing to mark **unresolved**, and a couple of entries in this KEP's Alternatives.

I hope that we can make this easy enough to teach for the common case of “I want to grant access to a thing that I name (eg Ingress / `secretName`)”, and still leave room during beta to make the more complex cases possible.

I also recommend sketching out the docs for how we explain this, because if that proves difficult, maybe the API isn't right yet.

baselineGrant: SameNamespace
```

Implementations of APIs can bundle ClusterReferenceConsumer resources as part of

> their deployment. This resource links an RBAC Subject (likely the ServiceAccount

Author: Updated, thanks!

used by the controller) to a ClusterReferencePattern. Once linked, controllers
will gain access to anything that has been granted following that pattern.

```yaml
kind: ClusterReferenceConsumer
apiVersion: reference.authorization.k8s.io/v1alpha1
metadata:
  name: contour-gateway
subject:
  kind: ServiceAccount
  name: contour
  namespace: contour-system
patternNames:
- gateway-tls
```

Finally, users can use ReferenceGrant to authorize specific references that
follow a ClusterReferencePattern. For example, the following ReferenceGrant

> authorizes references from Gateways in the `prod` namespace to the `acme-tls`

Reviewer 3: patternNames is the name of the ClusterReferencePattern.
The ClusterReferencePattern specifies group+resource.

> authorizes references from Gateways in the `prod` namespace to the `acme-tls`

Author: +1 to @stealthybox's reply here, but this KEP has actually moved away from patternNames based on recent review feedback so closing this out.

Secret in the `prod-tls` namespace:

```yaml
kind: ReferenceGrant
apiVersion: reference.authorization.k8s.io/v1alpha1
metadata:
  name: prod-gateways
  namespace: prod-tls

> patternName: gateway-tls

Reviewer: May be clearer:
```suggestion
patternName: gateway-tls # assume that there is a suitable ClusterReferencePattern
                         # (see earlier example)
```

Author: Yeah this was a common point of confusion, pushed up an update that removes the link to a pattern altogether here.

#### Gateway API Cross-Namespace Secret Reference

Authors of APIs can bundle ClusterReferenceGrants with their API definitions
for any references they expect to be common. For example, Gateway API could
bundle the following ClusterReferenceGrant to authorize same-namespace
references from Gateways to Secrets for "tls".

```yaml
kind: ClusterReferenceGrant
apiVersion: reference.authorization.k8s.io/v1alpha1
metadata:
  name: gateway-tls
from:
  group: gateway.networking.k8s.io
  resource: gateways
  version: v1
  path: ".spec.listeners[*].tls.certificateRefs[*]"
to:
  group: ""
  resource: secrets
for: tls
```

Implementations of APIs can bundle ClusterReferenceConsumer resources as part of
their deployment. This resource links a subject (likely the ServiceAccount
used by the controller) to matching (Cluster)ReferenceGrants. The subject of
this ClusterReferenceConsumer will receive authorization for any
(Cluster)ReferenceGrant with matching `from`, `to`, and `for` values.

```yaml
kind: ClusterReferenceConsumer
apiVersion: reference.authorization.k8s.io/v1alpha1
metadata:
  name: contour-gateway
subject:
  kind: ServiceAccount
  name: contour
  namespace: contour-system

> from:

Author: Do you mean that we should have a list that includes structs with from+to+for? Or do you mean that each of those fields should be lists? If the former, why not just create multiple instances of the resource?

> from:

Reviewer 3: Yeah, I was thinking a list that includes structs with from+to+for :+1:

> why not just create multiple instances of the resource?

It could go either way.
I'm inspired by RBAC which allows you to, within a single resource, list:
- resource+group+verb+names for a particular Role
- subjects for a particular RoleBinding

In this case, I see a Consumer as a single Subject:
- using multiple classNames
- referencing many different From+To+For

If it's not a list, I think you'd end up repeating Subject and classNames a lot for a complex API like Flux or Gateway.
Also, each object would probably also be named something like "{subject}-{from}-{to}-{for}".

I've updated the KEP to have `ClusterReferenceConsumer.references[].{from,to,for}` and added a Gateway example using GEP-91 for clientValidation to show how it looks.

  group: gateway.networking.k8s.io
  resource: gateways
to:

>   group: ""

Reviewer 5: Curious, why did we just include group and resource and not version here?

Reviewer 5: Ah, nvm, version isn't in rbac either

  resource: secrets
for: tls
```

Finally, users can use ReferenceGrant to authorize specific cross-namespace
references. For example, the following ReferenceGrant authorizes references from
Gateways in the `prod` namespace to the `acme-tls` Secret in the `prod-tls`
namespace:

```yaml
kind: ReferenceGrant
apiVersion: reference.authorization.k8s.io/v1alpha1

> metadata:

Reviewer 6: I think that it would be desirable to add "namespace: prod-tls" to the metadata section.

> metadata:

Reviewer 6: I think that it would be desirable to add "namespace: prod-tls" to the metadata section.

Author: Updated, thanks!

  name: prod-gateways
from:
  group: gateway.networking.k8s.io
  resource: gateways
  namespace: prod
to:
  group: ""
  resource: secrets
  names:
  - acme-tls
for: tls
```

### Existing ReferenceGrant Usage Examples

#### Gateway API HTTPRoute Referencing Service

In this example, a HTTPRoute in the `baz` namespace is directing traffic
to a Service backend in the `quux` namespace.

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:

>   name: quuxapp

Reviewer 2: ```suggestion
  name: quux-route
```

to make it easier to read, now the route and the service have the same name which makes it hard to read.

Author: Updated, thanks!

      namespace: quux
      port: 80

> ---

Reviewer 2: Could we show an example referenced service here as well for a full example?

    namespace: prod
  volumeMode: Filesystem
---
apiVersion: gateway.networking.k8s.io/v1beta1
kind: ReferenceGrant
metadata:
  name: allow-prod-pvc
  namespace: prod
spec:
  from:
  - kind: PersistentVolumeClaim
    namespace: dev
  to:
  - group: snapshot.storage.k8s.io

>     king: VolumeSnapshot

Reviewer 6: typo Kind

Author: Updated, thanks!

>     king: VolumeSnapshot

Reviewer 2: ```suggestion
    kind: VolumeSnapshot
```

Author: Updated, thanks!

### API Spec

#### ClusterReferenceConsumer

> 

Author: Completely agreed. I think I've got both in the KEP + POC is already in progress here: https://github.com/robscott/referencegrant-poc.

```golang
// ClusterReferenceConsumer identifies a common form of referencing pattern. This
// can then be used with ReferenceGrants to selectively allow references.
type ClusterReferenceConsumer struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// Subject refers to the subject that is a consumer of the referenced
	// pattern(s).

> 	Subject rbacv1.Subject `json:"subject"`

Reviewer 7: I encourage avoiding the RBAC type. I regret placing an APIGroup in the RBAC subject. I would exclude it in future APIs and knowing then what I know now, would have made a "normal" discriminated union.

Author: Good point, created a new type for this.

	// The names of the ClusterReferencePatterns this consumer implements.
	PatternNames []string `json:"patternNames"`
}
```

#### ClusterReferencePattern

> 

Author: Completely agreed, CRD examples are included below this + there's a POC.

```golang
// ClusterReferencePattern identifies a common form of referencing pattern. This
// can then be used with ReferenceGrants to selectively allow references.

> type ClusterReferencePattern struct {

Reviewer 7: some bits above indicate that `verb` should be included here.  Oversight here or mistake there?  Verb makes for interesting possibilities.

Author: Good catch, verb was a consideration in earlier versions of this KEP but was intentionally omitted in favor of this translating to `[get, list, watch]` for everything to start. It's possible we may make this configurable in the future. (cc @liggitt for another opinion here)

	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// Group is the group of the referent.
	Group string `json:"group"`

	// Resource is the resource of the referent.
	Resource string `json:"resource"`

	// Version is the API version of this resource this path applies to.
	Version string `json:"version,omitempty"`

	// Path is the path which this reference may come from.
	Path string `json:"path"`

	// BaselineGrant allows granting access to same-namespace references by
	// default without the need for ReferenceGrants.
	BaselineGrant string `json:"baselineGrant"`
}
```

#### ReferenceGrant

> 

Author: I do have an "Example Usage" section above this that includes YAML examples of each resource, is that sufficient here?

### API Spec

#### ClusterReferenceConsumer

```golang

> // Subject refers to the subject that is a consumer of the referenced
> // pattern(s).

Reviewer 2: ```suggestion
// ClusterReferenceConsumer authorizes the given subject to access a referenced resource
// in the same namespace.
```

Reviewer 2: Could probably word this better still.

Reviewer 3: Consumer resources do not authorize.
They opt-in to matching Grants.

These comments do need to be cleaned up, because it looks like they are stale.

type ClusterReferenceConsumer struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Subject refers to the subject that is a consumer of the referenced
    // pattern(s).
    Subject Subject       `json:"subject"`

    // From refers to the group and resource that these references originate from.

>     From    GroupResource `json:"from"`

Reviewer 2: Do you aim to add this struct to the API spec, or is it clearly just a struct of Group and Resource fields?

Author: I was hoping it was fairly clear, but don't mind adding it if it's helpful. (It's already in the POC fwiw).

Reviewer 2: Indeed, it's clear, but adding it doesn't hurt. Was just thinking as the other structs have more info than just GR, so just in case this one would as well.

    // To refers to the group and resource that these references target.
    To      GroupResource `json:"to"`

>     // For refers to the purpose of this reference. (Cluster)ReferenceGrants

Reviewer 2: I guess this string is case-sensitive? Shall we point that out?
Do we enforce this to e.g. be a DNS label, to avoid messiness with whitespace and similar?

Author: Good call, updated to clarify that it must be a DNS label as per RFC-1035 (would follow validation defined here: https://github.com/kubernetes/kubernetes/blob/11b974043604f5ccbeb6e5e62e1d9edcf00bc336/pkg/apis/apiserverinternal/validation/validation.go#L180-L181)

    // matching the From, To, and For of this resource will be authorized for
    // the Subject of this resource.
    For     string        `json:"for"`
}

// Subject is a copy of RBAC Subject that excludes APIGroup.
type Subject struct {

>     // Kind of object being referenced. Values defined by this API group are

Reviewer 2: Do we leave out `apiGroup` like in RBAC? https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.29/#subject-v1-rbac-authorization-k8s-io

I think we should. But just pointing it out as we already have this kind of struct elsewhere, and there it's relating the "User" and "Group" kinds with the "rbac.authorization.k8s.io" apiGroup (unfortunately, would indeed have been better to deem all those kinds as "core")

Reviewer 3: @liggitt and @deads2k recommended removing apiGroup here based off of learnings from RBAC

    // Name of the object being referenced.
    Name string `json:"name"`
    // Namespace of the referenced object.  If the object kind is non-namespace,
    // such as "User" or "Group", and this value is not empty the Authorizer
    // should report an error. +optional
    Namespace string `json:"namespace,omitempty"`
}
```

#### ClusterReferenceGrant

```golang
// ClusterReferenceGrant authorizes same-namespace references matching the
// specified pattern.
type ClusterReferenceGrant struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // From refers to the group, resource, and pthat that these references
    // originate from.
    From    GroupVersionResourcePath `json:"from"`

    // To refers to the group and resource that these references target.
    To      GroupResource `json:"to"`

    // For refers to the purpose of this reference. Subjects of
    // ClusterReferenceConsumers will be authorized to follow references matching
    // the From, To, and For of this resource.
    For     string        `json:"for"`
}

type GroupVersionResourcePath struct {
    Group    string `json:"group"`
    Resource string `json:"resource"`

>     Version  string `json:"version"`

Reviewer 2: Why do we specify version here? Just for being super safe on fully-qualifying everything, or is there a risk that `/apis/foo.com/v1/my-resource` and `/apis/foo.com/v2/my-resource` would serve completely different things, such that references would break?

In other places I see only GroupResource, why is `ClusterReferenceGrant.from` different?

Reviewer 3: The controller/authorizer has to watch the full contents of these resources to follow the Path to retrieve the names of each target resource we would like to granularly authorize.

Each Path is going to be specific to the structure of that versioned API.
The references would break between versions like you suggest.

We need to support multiple versions here.

A ClusterReferenceGrant could support multiple GroupVersionResourcePaths in a list.
The alternative is that multiple versions would have to be supported with multiple resources.

Author: +1 to what @stealthybox said, JSONPath could be interpreted differently depending on the API Version used.

Reviewer 2: True, thanks for pointing out that the path (of course) is version-specific, makes sense like this then.

Then the bigger question is that shall we support a list of GroupVersionResourcePaths or not? I think maybe yes, although it increases the complexity a little, as it's no fun having to manage multiple API objects on upgrades.

>     // Path is the path which this reference may come from.

Reviewer 2: Shall we specify here that this needs to be a valid JSONPath?
And/or actually name it `jsonPath` to be explicit, and allow e.g. CEL to be neatly added in the future.

I guess this field is optional, then we should specify that and add omitempty.

Is there any risk that users would like to have a list of strings instead?
Of course, they could make two objects, but it's convenient to make this a list if that's deemed to be a common use-case here.

Author: Good call, I've renamed this to JSONPath. I think it's unlikely that a user would want to specify multiple JSONPath, but if they did, they could always create multiple ClusterReferenceGrants.

### API Spec

#### ClusterReferenceConsumer

```golang
// Subject refers to the subject that is a consumer of the referenced
// pattern(s).
type ClusterReferenceConsumer struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // Subject refers to the subject that is a consumer of the referenced
    // pattern(s).
    Subject Subject       `json:"subject"`

    // From refers to the group and resource that these references originate from.
    From    GroupResource `json:"from"`

    // To refers to the group and resource that these references target.
    To      GroupResource `json:"to"`

    // For refers to the purpose of this reference. (Cluster)ReferenceGrants
    // matching the From, To, and For of this resource will be authorized for
    // the Subject of this resource.
    //

>     // This value must be a valid DNS label as defined per RFC-1035.

Reviewer 2: ```suggestion
    // This value must be a valid DNS label as defined per RFC-1035, and is matched case-sensitively.
```

    // recognized the kind value, the Authorizer should report an error.
    Kind string `json:"kind"`
    // Name of the object being referenced.
    Name string `json:"name"`
    // Namespace of the referenced object.  If the object kind is non-namespace,
    // such as "User" or "Group", and this value is not empty the Authorizer
    // should report an error. +optional
    Namespace string `json:"namespace,omitempty"`
}
```

#### ClusterReferenceGrant

```golang
// ClusterReferenceGrant authorizes same-namespace references matching the
// specified pattern.
type ClusterReferenceGrant struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // From refers to the group, resource, and pthat that these references
    // originate from.
    From    GroupVersionResourcePath `json:"from"`

    // To refers to the group and resource that these references target.
    To      GroupResource `json:"to"`

    // For refers to the purpose of this reference. Subjects of
    // ClusterReferenceConsumers will be authorized to follow references matching
    // the From, To, and For of this resource.
    //

>     // This value must be a valid DNS label as defined per RFC-1035.

Reviewer 2: ```suggestion
    // This value must be a valid DNS label as defined per RFC-1035, and is matched case-sensitively.
```

Reviewer: Can we have this as a case-insenstive match?

Reviewer 2: Yes, we can, I just want to clarify whether we have one or the other. As long as it's documented, I'm fine with either.

Author: The relevant RFC actually only supports lowercase characters, will update the KEP to clarify: https://github.com/kubernetes/kubernetes/blob/d2904845cdbbe16f418a3352ae4312775e657d24/pkg/apis/apiserverinternal/validation/validation.go#L180-L181

    For     string        `json:"for"`
}

type GroupVersionResourcePath struct {
    Group    string `json:"group"`
    Resource string `json:"resource"`
    Version  string `json:"version"`

>     // JSONPath is the JSONPath which this reference may come from.
>     JSONPath string `json:"jsonPath"`

Reviewer 2: ```suggestion
    // JSONPath is the JSONPath which this reference may come from.
    //+optional
    JSONPath string `json:"jsonPath,omitempty"`
```

The JSONPath should be optional, right?

Author: I can't think of any way to implement this without a jsonPath (or CEL path). This all relies on the authorizer being able to follow the path of a reference to authorize access to the resource that is targeted by the reference that this jsonPath points to.

Reviewer 2: Ok, thanks for the clarification 👍

}
```

#### ReferenceGrant

```golang
// ReferenceGrant identifies namespaces of resources that are trusted to
// reference the specified names of resources in the same namespace as the
// grant.
type ReferenceGrant struct {
    metav1.TypeMeta   `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`

    // From refers to the group and resource that these references originate from.
    From    GroupResource `json:"from"`

    // To refers to the group, resource, and names that these references can
    // target.
    To      ReferenceGrantTo `json:"to"`

    // For refers to the purpose of this reference. Subjects of
    // ClusterReferenceConsumers will be authorized to follow references
    // matching the From, To, and For of this resource.
    //

>     // This value must be a valid DNS label as defined per RFC-1035.

Reviewer 2: ```suggestion
    // This value must be a valid DNS label as defined per RFC-1035, and is matched case-sensitively.
```

Reviewer: Can we have this as a case-insenstive match?

  specific `To` Kinds.
* Corollary for future work, define how controllers interact. Is it a problem if
  multiple controllers consume the same ReferenceGrant?
* Status design is still pending, but it's currently expected that controllers
  will indicate status on the _referring_ resources, not on ReferenceGrant itself.
<<[UNRESOLVED do we need status? ]>>

> Original Thread: https://github.com/kubernetes/enhancements/pull/3767#discussion_r1084670421

Reviewer 2: Could we leave a little bit of explanation of this, why we don't want to put it on the ReferenceGrant side?

Author: Added a note below in "Do We Need Status?" section that replaced this, the key point is that it's difficult to handle status here because it's unclear who would populate it (many controllers involved).

> * Similar to the Gateway API version of ReferenceGrant, controllers will

Reviewer 2: Can we (un)officially "reserve" some status sub-field (like `allowedReferenceGrants`), such that one could optimistically query these referring resources' status in some generic way?

#### ClusterReferenceGrant

```golang
// ClusterReferenceGrant identifies a common form of referencing pattern. This
// can then be used with ReferenceGrants to selectively allow references.
type ClusterReferenceGrant struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// From refers to the group and resource that these references originate
	// from.
	From GroupResource `json:"from"`

	// Versions describes how references and class partitions are defined for
	// the "From" API. Each Version string must be unique.

> 	Versions []VersionedReferencePaths `json:"versions"`

Reviewer 4: IIUC, the upper list `Versions` is only used for different versions, so we dont end up repeating versions and classes. 
The inner list `References` is used to specify the path+to+for combinations. Like the example with `tls-client-validation` above.

### Do We Need Label Selectors?

#### Rationale
We are leaving this feature out because it dramatically increases the complexity
of these APIs (both for users and implementations). These APIs do allow
authorizing all same-namespace references, and support a list of names for
cross-namespace references.

#### Context
Previous Discussions:
* https://github.com/kubernetes/enhancements/pull/3767#discussion_r1084492070
* https://github.com/kubernetes/enhancements/pull/3767#discussion_r1084674648

As a natural next step, instead of simply allowing a named list of resources, we
could use label selectors to enable:

> 1. Access to resources with specific labels

Reviewer 2: I guess this essentially and practically would force us to build the labelSelector feature into RBAC?

Author: +1, I don't think that's really viable, or if it is, it could/should be done separately.

* Clarify that the expected operating model for implementations expects them to
  have broad, read access to both ReferenceGrant and the specific `To` Kinds they
  support, and then self-limit to only _use_ the relevant ones.
* Determine if strings used in `For` are completely separate from verbs.
* Decide if supporting verbs such as `[create, update, delete]` make sense.
* Decide if supporting virtual verbs such as `[forward, scale, dance]` make sense.

>   We already have `For` which has a similar function.

Reviewer 4: +1

* Decide whether to expose/hide the API based on the API server feature flag
  and/or authorizer config.
* Decide how cross-namespace ReferenceGrants join target resources to their
  originating reference. One way of accomplishing this is to extract the
  namespace alongside the name of a target reference when reconciling 
  ClusterReferenceGrants, then pair those with a matching ReferenceGrant. A 
  trade-off of this approach is that it makes ReferenceGrants
  dependent on ClusterReferenceGrants. The alternative is that the subject of a ClusterReferenceConsumer is authorized for the granted object regardless of 

>   whether there is actually a cross-namespace reference.

Reviewer: We can't assume that a Kubernetes cluster has RBAC authz active. Did you mean “rely on the authz layer”?

* Decide whether to formally add `*` as a special value for Namespace, to mean
  "all namespaces".
* Decide whether to formally add `*` as a special value for classNames, to mean
  "all classNames". Should access to a class-partitioned API be open or closed 
  for a ClusterReferenceConsumer that lists no classNames? Software authors, API
  owners, and cluster admins all own different pieces of the config here. An 
  optional field that fails open can easily be missed by a cluster admin. An  

>   opt-in wildcard can become canonical in practice.

Reviewer 4: but what about APIs without the Class concept?

>   opt-in wildcard can become canonical in practice.

Reviewer 3: I think Rob means he's convinced `classNames` are required for Consumers to receive authorization, only when API Authors have indicated that there is a `classPath` on their cluster-level Grants.
Basically if `classNames` is missing, it will fail closed as opposed to failing open and granting access.

In the case where the API is not split on `classPaths`, no `classNames` is required for any access.

###### Will enabling / using this feature result in introducing new API types?

API Type: ReferenceGrant

> Supported Number of Objects per Cluster: No limit

Reviewer 2: This made me smile, good that we are optimistic 😄 #nolimits

Author: Haha whoops, updated to add more realistic goals here.
